src/transaction/log_replication.cpp

```diff
@@ -1,173 +1,264 @@
+  {
+    m_reader.advance_when_does_not_fit (sizeof (T));
+    const T log_rec = m_reader.reinterpret_copy_and_add_align<T> ();
+
+    // To allow reads on the page server, make sure that all changes are visible.
+    // Having log_Gl.hdr.mvcc_next_id higher than all MVCCID's in the database is a requirement.
+    MVCCID mvccid = log_rv_get_log_rec_mvccid (log_rec);
+    if (mvccid != MVCCID_NULL && !MVCC_ID_PRECEDES (mvccid, log_Gl.hdr.mvcc_next_id))
+      {
+	log_Gl.hdr.mvcc_next_id = mvccid;
+	MVCCID_FORWARD (log_Gl.hdr.mvcc_next_id);
+      }
+
+    const VPID rec_vpid = log_rv_get_log_rec_vpid<T> (log_rec);
+    // at this point, vpid can either be valid or not
+    const LOG_DATA &rec_log_data = log_rv_get_log_rec_data<T> (log_rec);
+
+    const bool need_sync_redo = log_rv_need_sync_redo (rec_log_data.rcvindex);
+    assert (rec_log_data.rcvindex != RVDK_UNRESERVE_SECTORS || need_sync_redo);
+
+    if (m_parallel_replication_redo == nullptr || VPID_ISNULL (&rec_vpid) || need_sync_redo)
+      {
+	// invoke sync
+
+	// To apply RVDK_UNRESERVE_SECTORS, one must first wait for all changes in this sector to be redone.
+	// Otherwise, asynchronous jobs skip redoing changes in this sector's pages because they are seen as
+	// deallocated. When the same sector is reserved again, redo is resumed in the sector's pages, but
+	// the pages are not in a consistent state. The current workaround is to wait for all changes to be
+	// finished, including changes in the unreserved sector.
+	if (m_parallel_replication_redo != nullptr && rec_log_data.rcvindex == RVDK_UNRESERVE_SECTORS)
+	  {
+	    m_parallel_replication_redo->wait_for_idle ();
+	  }
+
+	log_rv_redo_record_sync<T> (&thread_entry, m_reader, log_rec, rec_vpid, rec_lsa,
+				    nullptr, rectype, m_undo_unzip, m_redo_unzip);
+      }
+    else
+      {
+	// dispatch async
+	using redo_job_impl_t = cublog::redo_job_impl<T>;
+	std::unique_ptr<redo_job_impl_t> job
+	{
+	  new redo_job_impl_t (rec_vpid, rec_lsa, nullptr, rectype, true)
+	};
+	m_parallel_replication_redo->add (std::move (job));
+      }
+  }
+
+  void
+  replicator::wait_replication_finish () const
+  {
+    std::unique_lock<std::mutex> ulock (m_redo_lsa_mutex);
+    m_redo_condvar.wait (ulock, [this]
+    {
+      return m_redo_lsa >= log_Gl.append.get_nxio_lsa ();
+    });
+
+    // at this moment, if using parallel replication, probably, MOST data has
+    // been dispatched for async replication;
+    // cannot decide if NO OTHER DATA will ever be fed because the daemon is still executing;
+    // will be able to tell that no data will ever be fed again for async replication
+    // only when the daemon is not running anymore;
+    // however, at this point introduce a fuzzy syncronization point by waiting all
+    // fed data to be effectively consumed/applied
+    if (m_parallel_replication_redo != nullptr)
+      {
+        m_parallel_replication_redo->wait_for_idle ();
+      }

```

Comment:
I do not like the semantics of this function.
I added the bits for the async tasks but there is nothing in place that stops the daemon/async-tasks infrastructure resuming the consumption/application of data.
Actually the only place where we can be sure that the consumption has stopped is in the dtor where the daemon is stopped and the async tasks are waited to terminate work.

Reply:
In theory, the function is called after all log records have been generated. nxio_lsa should not increase anymore, so if m_redo_lsa catches up, that's it, it is done. In this context, the current function is enough.

It is not meant to be called while log records are still being generated. Stronger guarantees would be possible, by forcing daemon somehow not pickup anymore records.

Or maybe it is better to make the function purpose more clear, by renaming as wait_replication_finish_during_shutdown(). Add a comment inside that nxio_lsa is no longer expected to grow.

Reply:
Done.